### ISSUE TITLE
Bot Framework Getting 502 Bad Gateway Errors in Production Environment after the Bot is not Requested for around 10 Minutes (Node.js)

### ISSUE THREAD

**Issue by PeterChen1105** [microsoft/botbuilder-js#1432]
When I use Microsoft Bot Framework for developing a conversational bot using multiple waterfall dialogs, I always receive **502 Bad Gateway** errors in the **Production Environment** when the bot is not requested for around **10 Minutes**.

However, in the local environment, the error will not happen when the application is tested using Bot Emulator.

I have been looking for solutions regarding this issue, and it seems that the error is caused by the bot not being able to send back the response within **15 Seconds**.

Here are the things that I have tried in order to get around with the issue:

- Enabled "Always on" options for Azure Cloud Application (cannot solve the issue)

<img width="465" alt="Azure  Alway On  Option Enabled" src="https://user-images.githubusercontent.com/53604754/69195973-44223280-0b81-11ea-86c6-4061df480441.png">

- Changed the in-memory storage of state into custom storage using Cosmos DB (cannot solve the issue)

<img width="627" alt="Cosmos DB for State Storage" src="https://user-images.githubusercontent.com/53604754/69195983-497f7d00-0b81-11ea-9aad-89cb031e4746.png">

- Sending Proactive Messages for certain long-running processes. However, when the 502 error starts to happen, even fetching one document from Firebase Firestore can be the "long-running process" which results in the error. Since my application requires checking Firebase document status constantly to see whether certain conversation stage is open and decide which **Dialog** should be applied, it is unlikely to identify long-running processes in this context. 

<img width="581" alt="Firebase Request in the Initial Dialog" src="https://user-images.githubusercontent.com/53604754/69195984-4d130400-0b81-11ea-80f1-893a6f0d6ef8.png">

Based on my observation, when the 502 gateway error occurs, the bot will be stuck in the initial stage rather than running slowly (i.e. the bot will be stuck in the initial Firebase request), and it does not seem to be the problem of long-running processes. It is not a problem with Firebase since fetching data from Firebase (which basically should take no time) will appear to be long-running processes when the bot does not appear to be "active" after around 10 minutes. The issue is that the bot seems not being capable of sending the fetching requests.

Also, when the error occurs, the error message throwing by the bot framework is as follows, and I cannot find any related solutions for the error `activity.code with value "14" must be of type string`.

>[onTurnError]: Error: No connection established
>BotFrameworkAdapter.processActivity(): 500 ERROR - Error: activity.code with value "14" must be of type string. (node:5132) UnhandledPromiseRejectionWarning: Error: Error: activity.code with value "14" must be of type string. At BotFrameworkAdapter.processActivity (D\home\site\wwwroot\node_modules\botbuilder\lib\botFrameworkAdapter.js:511:19) at process._tickCallback (internal/process/next_tick.js:68:7). (node:5132) Unhandled Promise Rejection Warning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled
with .catch(). (rejection id: 9)

Considering this situation, what is the proper workaround of the issue?

- I am trying to write a cloud function to actively sending requests to the bot at a 5 minutes interval to keep it alive, but I am not sure whether it will be working and it does not seem to be the proper solution.
- If I need to send proactive messages for every received message (since it is not possible to identify long-running processes), how can I push the waterfall dialog forward & prompt users for certain inputs with proactive messages? Since the Proactive Message Sample *https://docs.microsoft.com/en-au/azure/bot-service/bot-builder-howto-proactive-message?view=azure-bot-service-4.0&tabs=javascript* does not provide an implementation to continue the waterfall dialog & prompt user for certain inputs (e.g. AttachmentPrompt, NumberPrompt, etc.).

**Comment by jwiley84** [comment]
Looking into 
> activity.code with value "14" must be of type string

as that is not an error I recognize offhand.

**Comment by benbrown** [comment]
You said:

>  how can I push the waterfall dialog forward & prompt users for certain inputs with proactive messages

It is possible to create a dialog context and use it to begin the dialog _from within the proactive context_ created when you call `adapter.continueConversation`.

**Comment by PeterChen1105** [comment]
@benbrown Hi, could you please specify a little bit about? If I am trying to forward all the messages received to **proactive context**, what should I do in the following function? Just send a HTTP request to the /api/notify endpoint?

```
adapter.processActivity(req, res, async turnContext => {
    //
}
```

Also, in the /api/notify route, after I get the conversationReference, this function seems not to be working. Using `adapter.continueConversation`, I can only `await turnContext.sendActivity(msg)`, but I cannot run the current context. Could you provide some details regarding this?

```
await adapter.continueConversation(
      conversationReference,
      async turnContext => {
        return await bot.run(turnContext);
      }
    );
```

**Comment by PeterChen1105** [comment]
@jwiley84 Hi, is there any update for the error? I tried to activate the bot using a scheduler (sending the request to the bot at a five-minute interval in order to keep it alive), but it seems that it can only prolong the lifetime of the bot. And after 1 hour or so, I will still receive the same error message as `activity.code with value "14" must be of type string`.

And after receiving this error message, all the consequent messages will receive a 502 error.

**Comment by carlosscastro** [comment]
Hello @PeterChen1105! Our support team looked into this item and escalated it to the dev team. I'm investigating, and will come back with questions and / or checks you can do to get to the resolution of this item. I'll reach out after I catch up and do a first round of investigation here.

**Comment by PeterChen1105** [comment]
Hi @carlosscastro, thx for your information, cheers

**Comment by carlosscastro** [comment]
Hello @PeterChen1105!

There is a lot to unpack here so I'd like us to go back to the original issue. First, I'd like to note that if you could provide a minimal code example that demonstrates the problem with very little code and precise instructions, we could try it ourselves and analyze if there is a bug. So if you feel inclined and really want to get this fixed, I'd definitely recommend isolating the problem and how to get to it and sending us the steps and **minimal** code.

In the meantime, I'd like to organize the information here, particularly the **original** report:

- The bot is a conversational bot using multiple waterfall dialogs
- Customer always receives 502 Bad Gateway errors in the Production Environment when the bot is not requested for around 10 Minutes.
- On the local environment, the error will not happen when the application is tested using Bot Emulator.
- Enabling always or switching to memory store does not help

Then you mentioned that you tried proactive as a means to fixing this. Since it didn't fix it, potentially proactive would just complicate things further so I'd like to tackle the original issue.

Note that some of your symptoms make sense, since the timeout in directline when your bot is deployed to production is 15 seconds, while in the emulator we don't enforce that. This means that if your bot takes more than 15 seconds to respond we'll stop waiting.

Since I don't have a repro of your issue at this point, here are a few things that I'd recommend trying in order to gather more information. I cannot guarantee that any of these will lead to the root cause but they should help get closer to it.

1. Simplify and obtain minimal repro: As I mentioned below, remove all your custom logic and just replace with hardcoded answers in the waterfall and no database calls. Does the bot still get stuck? If yes, send us that code along with instructions on how to deploy and we can try it.
2. App insights: Are you using app insights? It's an amazing way of knowing what is going on with your bot, how long your bot is taking to respond, how long it's spending waiting for external calls to respond, etc
3. Deploy a sample: You could try getting a bot sample from botbuilder-samples and deploying it to see if it gets stuck for you: https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/javascript_nodejs/16.proactive-messages and then add your logic to it to see if that makes it get stuck

If I were in your position, I'd go for 2. and 1. If in 1 you remove your custom logic and the bot is not stuck anymore, you can tell which was the offending line and find a way to fix it.

**Comment by cleemullins** [comment]
@PeterChen1105 - Hi Peter, are you able to update this thread with a current status?

**Comment by carlosscastro** [comment]
Hi @PeterChen1105, just checking in if you were to get a small repro of the issue so we can help you debug.

**Comment by PeterChen1105** [comment]
Hi @carlosscastro @cleemullins I have contacted the Microsoft Support Team to get some extra help, and it seems that the error is caused by Firestore that I used for persisting data and push real-time notifications.

`
Thu Dec 12 2019 15:14:56 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 1, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 16:14:56 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 1, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 17:14:58 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 1, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 17:14:58 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 2, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 17:14:58 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 3, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:00 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 1, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:00 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 2, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:01 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 3, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:03 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 4, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:05 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 5, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:09 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 6, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:11 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 7, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:16 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 8, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:22 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 9, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:35 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 10, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:15:58 GMT+0000 (Greenwich Mean Time): enter backoffAndWait, retryCount+1, this.retryCount: 11, exports.MAX_RETRY_ATTEMPTS: 10
Thu Dec 12 2019 18:16:34 GMT+0000 (Greenwich Mean Time): enter backoffAndWait error situation, this.retryCount: 11, exports.MAX_RETRY_ATTEMPTS: 10
Encountered error: Error: Exceeded maximum number of retries allowed
`

When encountering a maximum number of retries, the next bot message will receive a timeout error.

I tried to update the version of Firestore and started monitoring the bot status from the day before yesterday, and up until now, the error does not occur for 24 hours. So I will keep monitoring the status and hopefully, the problem has been tackled by the latest version of Firestore.

Sorry for the late update of the thread status since to reproduce the error, sometimes it will take more than 10 hours.
